### PR TITLE
namespace_sync_controller: buildWorks parallel sync namespace

### PR DESCRIPTION
**Pull request by helen-frank**
Signed-off-by: helen <haitao.zhang@daocloud.io>

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
accelerate namespace sync
**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/controllers/namespace/namespace_sync_controller.go [hunk 1]
```diff
@@ -107,47 +108,59 @@ func (c *Controller) buildWorks(namespace *corev1.Namespace, clusters []clusterv
 		return err
 	}
 
-	for _, cluster := range clusters {
-		clonedNamespaced := namespaceObj.DeepCopy()
+	errChan := make(chan error, len(clusters))
+	for i := range clusters {
+		go func(cluster *clusterv1alpha1.Cluster, ch chan error) {
```
Comment (XiShanYongYe-Chang) [inline pkg/controllers/namespace/namespace_sync_controller.go:113]: How about limiting the `ch` usage here?

```
ch chan<- error
```

File: pkg/controllers/namespace/namespace_sync_controller.go [hunk 2]
```diff
@@ -107,47 +108,59 @@ func (c *Controller) buildWorks(namespace *corev1.Namespace, clusters []clusterv
 		return err
 	}
 
-	for _, cluster := range clusters {
-		clonedNamespaced := namespaceObj.DeepCopy()
+	errChan := make(chan error, len(clusters))
```
Comment (XiShanYongYe-Chang) [inline pkg/controllers/namespace/namespace_sync_controller.go:111]: Do we really need to define the length of the channel? I think it's okay to use unbuffered channels.

Comment (helen-frank) [reply pkg/controllers/namespace/namespace_sync_controller.go:111]: My consideration is that when there are too many ns, unbuffered pipe blocking may cause goroutine to hold too much

Comment (XiShanYongYe-Chang) [reply pkg/controllers/namespace/namespace_sync_controller.go:111]: ok.

In addition, perhaps we could add the appropriate namespace information when handling the error information, indicating which namespaces were in error.

Comment (helen-frank) [reply pkg/controllers/namespace/namespace_sync_controller.go:111]: ok, wait a minute

Comment (helen-frank) [reply pkg/controllers/namespace/namespace_sync_controller.go:112]:
```go
fmt.Errorf("%v sync namespace %s failed, err: %v", cluster.GetName(), clonedNamespaced.GetName(), err)
```
What do you think

Comment (XiShanYongYe-Chang) [reply pkg/controllers/namespace/namespace_sync_controller.go:112]: Looks good to me~

File: pkg/controllers/namespace/namespace_sync_controller.go (hunk 1 above)
Comment (XiShanYongYe-Chang) [reply pkg/controllers/namespace/namespace_sync_controller.go:114]: What about this question?

Comment (helen-frank) [reply pkg/controllers/namespace/namespace_sync_controller.go]: fixed

**Review by XiShanYongYe-Chang — COMMENTED** [review]
/lgtm
/assign @RainbowMango

File: pkg/controllers/namespace/namespace_sync_controller.go [hunk 3]
```diff
@@ -107,47 +109,59 @@ func (c *Controller) buildWorks(namespace *corev1.Namespace, clusters []clusterv
...
-		cops, _, err := c.OverrideManager.ApplyOverridePolicies(clonedNamespaced, cluster.Name)
-		if err != nil {
-			klog.Errorf("Failed to apply overrides for %s/%s/%s, err is: %v", clonedNamespaced.GetKind(), clonedNamespaced.GetNamespace(), clonedNamespaced.GetName(), err)
-			return err
-		}
+			// namespace only care about ClusterOverridePolicy
+			cops, _, err := c.OverrideManager.ApplyOverridePolicies(clonedNamespaced, cluster.Name)
+			if err != nil {
+				klog.Errorf("Failed to apply overrides for %s/%s/%s, err is: %v", clonedNamespaced.GetKind(), clonedNamespaced.GetNamespace(), clonedNamespaced.GetName(), err)
+				ch <- fmt.Errorf("%v sync namespace %s failed, err: %v", cluster.GetName(), clonedNamespaced.GetName(), err)
```
Comment (RainbowMango) [inline pkg/controllers/namespace/namespace_sync_controller.go:121]:
```suggestion
				ch <- fmt.Errorf("sync namespace(%s) to cluster(%s) failed due to: %v", clonedNamespaced.GetName(), cluster.GetName(), err)
```
Same as below.

**Comment by RainbowMango** [comment]
Just one nit, otherwise LGTM.

I wonder have you run some tests, how much performance has been improved?

**Comment by helen-frank** [comment]
> Just one nit, otherwise LGTM.
> 
> I wonder have you run some tests, how much performance has been improved?

I don't have so many clusters to test at present. Logically, the most time-consuming place lies in network connection. The previous time-consuming calculation formula was (n1 n2 n3 n4), and now the time-consuming calculation formula is max(n1,n2,n3,n4), which reduces the time complexity from O(n) to O(1)

**Comment by helen-frank** [comment]
If you need benchmark data, you may have to wait for the weekend.

**Review by RainbowMango — APPROVED** [review]
/approve

Thanks for your explanation, I think the benchmark should be a little bit more complex, we need to stub a lot of stuff. I believe this would work.

**Comment by RainbowMango** [comment]
Leave LGTM to @XiShanYongYe-Chang 
for the `force-push`, I don't know if any code changed since your last review.

**Comment by XiShanYongYe-Chang** [comment]
Thanks~
/lgtm

**Comment by helen-frank** [comment]
I tried to use fake, but  ApplyOverridePolicies `err is: clusters.cluster.karmada.io "c-2" not found`, haha
I will try to learn kwok and simulate clusters to provide benchmark.
In addition, I found a bug in my code. There was no return when some err was generated. I'm very sorry. https://github.com/karmada-io/karmada/pull/2942
@RainbowMango

**Comment by RainbowMango** [comment]
no worries.

**Pull request merged**